platforms/core-configuration/model-core/src/main/java/org/gradle/internal/reflect/annotations/impl/AbstractHasAnnotationMetadata.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.reflect.annotations.impl;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.reflect.TypeToken;
+import org.gradle.internal.Cast;
+import org.gradle.internal.reflect.annotations.AnnotationCategory;
+import org.gradle.internal.reflect.annotations.HasAnnotationMetadata;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Optional;
+
+public class AbstractHasAnnotationMetadata implements HasAnnotationMetadata {
+    protected final Method method;
+    protected final TypeToken<?> declaredType;
+    protected final ImmutableMap<AnnotationCategory, Annotation> annotationsByCategory;
+    protected final ImmutableMap<Class<? extends Annotation>, Annotation> annotationsByType;
+
+    public AbstractHasAnnotationMetadata(Method method, ImmutableMap<AnnotationCategory, Annotation> annotationsByCategory) {
+        this.method = method;
+        method.setAccessible(true);
```

Comment:
It's a good question.  I just moved this from where it was already in `PropertyAnnotationMetadata`.  I suspect it could be handled differently, but I'm inclined to leave it as is for now and do something about it as follow up work.